File: backend/src/test/java/fi/aalto/ekanban/services/GameServiceSpec.java
package fi.aalto.ekanban.services;

import static com.greghaskins.spectrum.Spectrum.*;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.notNullValue;

import java.util.function.Supplier;

import com.greghaskins.spectrum.Spectrum;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.junit.runner.RunWith;
import org.mockito.Mockito;

import fi.aalto.ekanban.builders.GameBuilder;
import fi.aalto.ekanban.enums.GameDifficulty;
import fi.aalto.ekanban.models.db.games.Game;
import fi.aalto.ekanban.repositories.GameRepository;

@RunWith(Spectrum.class)
public class GameServiceSpec {

    private static GameService gameService;
    private static GameInitService gameInitService;
    private static PlayerService playerService;
    private static GameOptionService gameOptionService;
    private static GameRepository gameRepository;

    {
        describe("GameService", () -> {

            beforeAll(() -> {
                gameInitService = Mockito.mock(GameInitService.class);
                playerService = Mockito.mock(PlayerService.class);
                gameOptionService = Mockito.mock(GameOptionService.class);
                gameRepository = Mockito.mock(GameRepository.class);
                gameService = new GameService(gameInitService, playerService,
                                              gameOptionService, gameRepository);
            });

    describe("startGame", () -> {

        final Supplier<GameDifficulty> gameDifficulty = let(() -> GameDifficulty.NORMAL);
        final Supplier<String> playerName = let(() -> "Player");
        final Supplier<Game> newGame = let(() -> GameBuilder.aGame()
                                                    .withPlayerName(playerName.get())
                                                    .withDifficultyLevel(gameDifficulty.get())
                                                    .build());

        describe("with normal difficulty", () -> {

            beforeEach(() -> {
                Mockito.when(gameInitService.getInitializedGame(
                        Mockito.any(GameDifficulty.class),
                        Mockito.any(String.class))).thenReturn(newGame.get());
                Mockito.when(gameRepository.save(Mockito.any(Game.class))).thenReturn(newGame.get());
            });

            final Supplier<Game> createdGame = let(() ->
                    gameService.startGame(playerName.get(), gameDifficulty.get()));

            it("should create game with the given name", () -> {
                assertThat(createdGame.get().getPlayerName(), equalTo(playerName.get()));
            });

            it("should create game with the given game difficulty", () -> {
                assertThat(createdGame.get().getDifficultyLevel(), equalTo(gameDifficulty.get()));
            });

        });
    });

            describe("throwError", () -> {
                final Supplier<Game> newGame = let(() -> gameService.throwError());

                it("should throw runTimeException", () -> {
                    //thrown.expect(RuntimeException.class);
                    //newGame.get();
                });
            });

        });
    }

}

File: backend/src/spec/unit/game_service_spec.rb
require 'java'
require_relative '../spec_helper'

java_import 'org.mockito.Mockito'
java_import 'org.junit.rules.ExpectedException'

java_import 'fi.aalto.ekanban.builders.GameBuilder'
java_import 'fi.aalto.ekanban.services.GameService'
java_import 'fi.aalto.ekanban.services.PlayerService'
java_import 'fi.aalto.ekanban.services.GameOptionService'
java_import 'fi.aalto.ekanban.services.GameInitService'
java_import 'fi.aalto.ekanban.repositories.GameRepository'
java_import 'fi.aalto.ekanban.enums.GameDifficulty'
java_import 'fi.aalto.ekanban.models.db.games.Game'

describe GameService do

  before(:all) do
    @gameInitService = Mockito.mock(GameInitService.java_class)
    @playerService = Mockito.mock(PlayerService.java_class)
    @gameOptionService = Mockito.mock(GameOptionService.java_class)
    @gameRepository = Mockito.mock(GameRepository.java_class)
    @game_service = GameService.new(@gameInitService, @playerService,
                                    @gameOptionService, @gameRepository)
  end

  describe 'startGame()' do
    let(:player_name) {"Player"}
    let(:game_difficulty) {GameDifficulty::NORMAL}
    let(:new_game) {GameBuilder.aGame().
                        with_player_name(player_name).
                        with_difficulty_level(game_difficulty).
                        build()}

    context 'with normal difficulty' do

      before(:each) do
        Mockito.when(@gameInitService.getInitializedGame(
            Mockito.any(GameDifficulty.java_class), Mockito.anyString)).thenReturn(new_game)
        Mockito.when(@gameRepository.save(new_game)).thenReturn(new_game)
      end

      subject(:created_game) { @game_service.start_game(player_name, game_difficulty) }

      it 'should create game with the given name' do
        expect(created_game.player_name).to eq player_name
      end

      it 'should create game with the given game difficulty' do
        expect(created_game.difficulty_level).to eq game_difficulty
      end
      
    end
  end

end